src/baldr/incident_singleton.h

```diff
@@ -0,0 +1,306 @@
+   * @param state     the state to update
+   * @param tile_id   the tile id we are loading
+   * @param path      the path to the tile file
+   * @return the shared_ptr that owns the tile memory
+   */
+  static bool update_tile(const std::shared_ptr<state_t>& state,
+                          const valhalla::baldr::GraphId& tile_id,
+                          const std::string& path) {
+    // load the tile
+    auto tile = read_tile(path);
+    // see if we have a slot
+    auto found = state->cache.find(tile_id);
+    // if we dont have a slot make one, this must be synchronized
+    if (found == state->cache.cend()) {
+      // this shouldnt happen in lock free mode but can if you put unexpected tiles in the log/dir
+      if (state->lock_free.load()) {
+        LOG_WARN("Incident watcher skipped " + std::to_string(tile_id) +
+                 " because it was not found in the configured tile extract");
+        return false;
+      }
+      state->mutex.lock();
+      found = state->cache.insert({tile_id, {}}).first;
+      state->mutex.unlock();
+    }
+    // atomically store the tile shared_ptr, could be actually nullptr when there are no incidents
+    std::atomic_store_explicit(&found->second, tile, std::memory_order_release);
+    LOG_DEBUG("Incident watcher " + std::string(tile ? "loaded " : "unloaded ") +
+              std::to_string(tile_id));
+    return true;
+  }
+
+  /**
+   * Thread work function that continually checks for updates to incident tiles. The updates come in
+   * the form of timestamps in a memory mapped file of 64bit integers (one per tile). The first 25
+   * bits are the tile id and level (as a normal graphid) the remaining 39 bits are the timestamp.
+   * When the timestamp for the last check is older than a timestamp for a given file that file is
+   * replaced with whatever its contents are on disk.
+   *
+   * The thread begins by deciding whether its just scaning the directory (works for a small number of
+   * incidents) or using a memory mapped changelog to communicate about which incidents chnaged last.
+   * If the latter is used we preallocate
+   * @param config
+   * @param tileset
+   * @param state
+   */
+  static void watch(boost::property_tree::ptree config,
+                    std::unordered_set<valhalla::baldr::GraphId> tileset,
+                    std::shared_ptr<state_t> state) {
+    LOG_INFO("Incident watcher started");
+    // try to configure for changelog mode
+    std::unique_ptr<valhalla::midgard::sequence<uint64_t>> changelog;
+    filesystem::path inc_log_path(config.get<std::string>("incident_log", ""));
+    filesystem::path inc_dir;
+    try {
+      changelog.reset(new decltype(changelog)::element_type(inc_log_path.string()));
+    } // check for a directory scan mode config
+    catch (...) {
+      inc_dir = filesystem::path(config.get<std::string>("incident_dir", ""));
+      if (!filesystem::is_directory(inc_dir)) {
+        inc_dir = {};
```

Comment:
Maybe log a message that it uses memmap or directory scan? Could be useful in debugging.